src/front/glsl/preprocess.rs

```diff
@@ -0,0 +1,145 @@
+        let mut processed_string: String;
+        loop {
+            if inside_comment {
+                if let Some(pos) = processed.find("*/") {
+                    processed = processed.split_at(pos + 2).1;
+                    inside_comment = false;
+                    self.inside_comment = false;
+                    continue;
+                }
+            } else if let Some(pos) = processed.find("/*") {
+                if let Some(end_pos) = processed[pos + 2..].find("*/") {
+                    // comment ends during this line
+                    processed_string = processed.to_string();
+                    processed_string.replace_range(pos..pos + end_pos + 4, "");
+                    processed = &processed_string;
+                } else {
+                    processed = processed.split_at(pos).0;
+                    inside_comment = true;
+                }
+                continue;
+            }
+            break;
+        }
+        // strip leading whitespace
+        processed = processed.trim_start();
+        if processed.starts_with('#') && !self.inside_comment {
+            let mut iter = processed[1..]
+                .trim_start()
+                .splitn(2, |c: char| c.is_whitespace());
+            if let Some(directive) = iter.next() {
+                skip = true;
+                in_preprocess = true;
+                match directive {
+                    "version" => {
+                        skip = false;
+                    }
+                    "define" => {
+                        if let Some(rest) = iter.next() {
+                            let pos = rest
+                                .find(|c: char| !c.is_ascii_alphanumeric() && c != '_' && c != '(')
+                                .unwrap_or_else(|| rest.len());
+                            let (key, mut value) = rest.split_at(pos);
+                            value = value.trim();
+                            self.defines.insert(key.into(), self.subst_defines(value));
+                        }
+                    }
+                    "undef" => {
+                        if let Some(rest) = iter.next() {
+                            let key = rest.trim();
+                            self.defines.remove(key);
+                        }
+                    }
+                    "ifdef" => {
+                        if let Some(rest) = iter.next() {
+                            let key = rest.trim();
+                            self.if_stack.push((self.defines.contains_key(key), false));
+                        }
+                    }
+                    "ifndef" => {
+                        if let Some(rest) = iter.next() {
+                            let key = rest.trim();
+                            self.if_stack.push((!self.defines.contains_key(key), false));
+                        }
+                    }
+                    "else" => {
+                        if let Some(if_state) = self.if_stack.last_mut() {
```

Comment:
Another way to right this would be `let if_state = self.if_stack.last_mut().ok_or(Error::UnmatchedElse)?;`
Similar technique can be used in other places too

Reply:
think I'm starting to love `.ok_or()` 😄